api/envoy/type/v3/transformations.proto

```diff
@@ -0,0 +1,63 @@
+syntax = "proto3";
+
+package envoy.type.v3;
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.type.v3";
+option java_outer_classname = "TransformationsProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Text Transformations API]
+
+// TextTransformations define an API to apply a sequence of operations that can
+// be used to alter text before it is used for matching or routing. Multiple
+// actions can be applied in the same Transformation, forming a sequential
+// pipeline. The transformations will be performed in the order that they
+// appear.
+//
+// This API is a work in progress.
+
+message TextTransformation {
```

Comment:
Is this message name a bit too generic? Should perhaps this be in some HTTP specific namespace and named more specifically for HTTP normalization?

Reply:
@yangminzhu What do you think? StringMatcher should be able to consume it to normalize/transform on matching. These are mostly, but not completely, header values.

Reply:
I proposed the name `TextTransformation` for the purpose that it can be used as a generic text transformation API, which could be used for things more than just HTTP normalization (e.g. remove prefix, base64Decode, etc.) and embed directly in StringMatcher, etc.

Reply:
I don't feel very strongly about it either way. @envoyproxy/api-shepherds any opinions?

Reply:
Why are we putting HTTP specific things in a generic `TextTransformation`? I think just having an HTTP specific `PathTransformation` would make sense, and distinct messages for other things (e.g. normalizing headers, etc.)

Reply:
I'm going to make it distinct for each. I will unwrap it one layer so a `PathTransformation` is oneof a path-relevant operation kind. I can consume `repeated PathTransformation` in HCM, and you can consume a `PathTransformation` as one of a `TextTransformation` in `StringMatcher`. We can also add mutual types to our messages from a common place.

Reply:
sorry to re-open the discussion here, @htuch @asraa 

I think `TextTransformation` is a general class that can be used in many places, here at HTTP is just one use case, and path is just one use case, it could be host, HTTP header, some claim in JWT token or even a field in the dynamic metadata.

Not every transformation is needed for every of these cases, it's up to the user to choose which one they want to use.

I'm just a bit worried that making it `PathTransformation` is too restrictive and makes it hard to be reused in all other cases (e.g. host, header and etc.)

Reply:
We can add other transformations as well for these things. We don't want to reuse RFC3986 or merge slashes. I think having specialized messages for these distinct normalization is a good thing, e.g. `PathTransformation`, `HeaderKeyTransformation`, `HeaderValueTransformation`. Agree with @mattklein123 on namespace.

LGTM modulo Matt's comment ,will hand it over to @mattklein123 for merge.